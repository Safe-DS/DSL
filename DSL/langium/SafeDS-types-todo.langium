interface SdsAnnotationCall extends SdsAbstractCall {
	annotation?: @SdsAnnotation
}

interface SdsAnnotation extends SdsAbstractCallable, SdsAbstractCompilationUnitMember {
	constraint?: SdsConstraint
}

interface SdsConstraint extends SdsAbstractObject {
	body?: SdsBlock
}

interface SdsPipeline extends SdsAbstractCompilationUnitMember {
	body?: SdsBlock
}

interface SdsAttribute extends SdsAbstractClassMember, SdsAbstractProtocolToken {
	static: boolean
	^type?: SdsAbstractType
}

interface SdsSchemaPlaceholder extends SdsAbstractAssignee, SdsAbstractNamedTypeDeclaration {
}

interface SdsSchemaYield extends SdsAbstractAssignee {
	^type?: SdsSchemaType
}
