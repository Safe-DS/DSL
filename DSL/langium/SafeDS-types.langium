
interface SdsCompilationUnit extends SdsAbstractDeclaration {
	imports: SdsImport[]
	members: SdsAbstractAnnotatedObject[]
}

interface SdsAbstractDeclaration extends SdsAbstractAnnotatedObject {
	annotationCallList?: SdsAnnotationCallList
	name?: string
}

interface SdsAbstractAnnotatedObject extends SdsAbstractObject {
	annotationCalls: SdsAnnotationCall[]
}

interface SdsAbstractObject {
}

interface SdsAnnotationCall extends SdsAbstractCall {
	annotation?: @SdsAnnotation
}

interface SdsAbstractCall extends SdsAbstractObject {
	argumentList?: SdsArgumentList
}

interface SdsArgumentList extends SdsAbstractObject {
	arguments: SdsArgument[]
}

interface SdsArgument extends SdsAbstractExpression {
	parameter?: @SdsParameter
	value?: SdsAbstractExpression
}

interface SdsAbstractExpression extends SdsAbstractObject {
}

interface SdsParameter extends SdsAbstractLocalVariable {
	defaultValue?: SdsAbstractExpression
	^type?: SdsAbstractType
	variadic: boolean
}

interface SdsAbstractLocalVariable extends SdsAbstractDeclaration {
}

interface SdsAbstractType extends SdsAbstractObject {
}

interface SdsAnnotation extends SdsAbstractCallable, SdsAbstractCompilationUnitMember {
	constraint?: SdsConstraint
}

interface SdsAbstractCallable extends SdsAbstractObject {
	parameterList?: SdsParameterList
}

interface SdsParameterList extends SdsAbstractObject {
	parameters: SdsParameter[]
}

interface SdsAbstractCompilationUnitMember extends SdsAbstractDeclaration {
}

interface SdsConstraint extends SdsAbstractObject {
	body?: SdsBlock
}

interface SdsBlock extends SdsAbstractObject {
	statements: SdsAbstractStatement[]
}

interface SdsAbstractStatement extends SdsAbstractObject {
}

interface SdsAnnotationCallList extends SdsAbstractAnnotatedObject {
}

interface SdsImport extends SdsAbstractObject {
	importedNamespace?: string
	alias?: SdsImportAlias
}

interface SdsImportAlias extends SdsAbstractObject {
	name?: string
}

interface SdsClass extends SdsAbstractCallable, SdsAbstractClassMember, SdsAbstractCompilationUnitMember, SdsAbstractNamedTypeDeclaration {
	typeParameterList?: SdsTypeParameterList
	parentTypeList?: SdsParentTypeList
	body?: SdsClassBody
}

interface SdsAbstractClassMember extends SdsAbstractDeclaration {
}

interface SdsAbstractNamedTypeDeclaration extends SdsAbstractDeclaration {
}

interface SdsTypeParameterList extends SdsAbstractObject {
	typeParameters: SdsTypeParameter[]
}

interface SdsTypeParameter extends SdsAbstractNamedTypeDeclaration {
	variance?: string
	kind?: string
}

interface SdsParentTypeList extends SdsAbstractObject {
	parentTypes: SdsAbstractType[]
}

interface SdsClassBody extends SdsAbstractObject {
	members: SdsAbstractObject[]
}

interface SdsEnum extends SdsAbstractNamedTypeDeclaration, SdsAbstractClassMember, SdsAbstractCompilationUnitMember {
	body?: SdsEnumBody
}

interface SdsEnumBody extends SdsAbstractObject {
	variants: SdsEnumVariant[]
}

interface SdsEnumVariant extends SdsAbstractCallable, SdsAbstractNamedTypeDeclaration {
	typeParameterList?: SdsTypeParameterList
	constraint?: SdsConstraint
}

interface SdsFunction extends SdsAbstractCallable, SdsAbstractClassMember, SdsAbstractCompilationUnitMember, SdsAbstractProtocolToken {
	static: boolean
	typeParameterList?: SdsTypeParameterList
	resultList?: SdsResultList
	body?: SdsFunctionBody
}

interface SdsAbstractProtocolToken extends SdsAbstractDeclaration {
}

interface SdsResultList extends SdsAbstractObject {
	results: SdsResult[]
}

interface SdsResult extends SdsAbstractResult {
	^type?: SdsAbstractType
}

interface SdsAbstractResult extends SdsAbstractDeclaration {
}

interface SdsFunctionBody extends SdsAbstractObject {
	statements: SdsAbstractObject[]
}

interface SdsPredicate extends SdsAbstractCompilationUnitMember, SdsAbstractCallable {
	typeParameterList?: SdsTypeParameterList
	resultList?: SdsResultList
	body?: SdsBlock
}

interface SdsSchema extends SdsAbstractCompilationUnitMember {
	columnList?: SdsColumnList
}

interface SdsColumnList extends SdsAbstractObject {
	columns: SdsColumn[]
}

interface SdsColumn extends SdsAbstractObject {
	columnName?: SdsString
	columnType?: SdsAbstractType
}

interface SdsString extends SdsAbstractLiteral {
	value?: string
}

interface SdsAbstractLiteral extends SdsAbstractExpression {
}

interface SdsPipeline extends SdsAbstractCompilationUnitMember {
	body?: SdsBlock
}

interface SdsStep extends SdsAbstractCallable, SdsAbstractCompilationUnitMember {
	visibility?: string
	resultList?: SdsResultList
	body?: SdsBlock
}

interface SdsMemberType extends SdsAbstractType {
	member?: SdsNamedType
	receiver?: SdsAbstractType
}

interface SdsNamedType extends SdsAbstractType {
	declaration?: @SdsAbstractNamedTypeDeclaration
	nullable: boolean
	typeArgumentList?: SdsTypeArgumentList
}

interface SdsTypeArgumentList extends SdsAbstractObject {
	typeArguments: SdsTypeArgument[]
}

interface SdsTypeArgument extends SdsAbstractObject {
	typeParameter?: @SdsTypeParameter
	value?: SdsAbstractTypeArgumentValue
}

interface SdsAbstractTypeArgumentValue extends SdsAbstractObject {
}

interface SdsParenthesizedType extends SdsAbstractType {
	^type?: SdsAbstractType
}

interface SdsAttribute extends SdsAbstractClassMember, SdsAbstractProtocolToken {
	static: boolean
	^type?: SdsAbstractType
}

interface SdsProtocol extends SdsAbstractObject {
	body?: SdsProtocolBody
}

interface SdsProtocolBody extends SdsAbstractObject {
	subtermList?: SdsProtocolSubtermList
	term?: SdsAbstractProtocolTerm
}

interface SdsProtocolSubtermList extends SdsAbstractObject {
	subterms: SdsProtocolSubterm[]
}

interface SdsProtocolSubterm extends SdsAbstractProtocolToken {
	term?: SdsAbstractProtocolTerm
}

interface SdsAbstractProtocolTerm extends SdsAbstractObject {
}

interface SdsProtocolAlternative extends SdsAbstractProtocolTerm {
	terms: SdsAbstractProtocolTerm[]
}

interface SdsProtocolSequence extends SdsAbstractProtocolTerm {
	terms: SdsAbstractProtocolTerm[]
}

interface SdsProtocolQuantifiedTerm extends SdsAbstractProtocolTerm {
	term?: SdsAbstractProtocolTerm
	quantifier?: string
}

interface SdsProtocolComplement extends SdsAbstractProtocolTerm {
	universe?: SdsProtocolTokenClass
	referenceList?: SdsProtocolReferenceList
}

interface SdsProtocolTokenClass extends SdsAbstractProtocolTerm {
	value?: string
}

interface SdsProtocolReferenceList extends SdsAbstractProtocolTerm {
	references: SdsProtocolReference[]
}

interface SdsProtocolReference extends SdsAbstractProtocolTerm {
	token?: @SdsAbstractProtocolToken
}

interface SdsProtocolParenthesizedTerm extends SdsAbstractProtocolTerm {
	term?: SdsAbstractProtocolTerm
}

interface SdsLambdaParameterList extends SdsAbstractExpression, SdsParameterList {
}

interface SdsAssignment extends SdsAbstractStatement {
	assigneeList?: SdsAssigneeList
	expression?: SdsAbstractExpression
}

interface SdsAssigneeList extends SdsAbstractObject {
	assignees: SdsAbstractAssignee[]
}

interface SdsAbstractAssignee extends SdsAbstractObject {
}

interface SdsPlaceholder extends SdsAbstractAssignee, SdsAbstractLocalVariable {
}

interface SdsWildcard extends SdsAbstractAssignee {
}

interface SdsYield extends SdsAbstractAssignee {
	result?: @SdsResult
}

interface SdsExpressionStatement extends SdsAbstractStatement {
	expression?: SdsAbstractExpression
}

interface SdsSchemaPlaceholder extends SdsAbstractAssignee, SdsAbstractNamedTypeDeclaration {
}

interface SdsSchemaYield extends SdsAbstractAssignee {
	^type?: SdsSchemaType
}

interface SdsSchemaType extends SdsAbstractType {
	declaration?: @SdsAbstractNamedTypeDeclaration
}

interface SdsBlockLambda extends SdsAbstractLambda {
	body?: SdsBlock
}

interface SdsAbstractLambda extends SdsAbstractCallable, SdsAbstractExpression {
}

interface SdsExpressionLambda extends SdsAbstractLambda {
	result?: SdsAbstractExpression
}

interface SdsBlockLambdaResult extends SdsAbstractAssignee, SdsAbstractResult {
}

interface SdsInfixOperation extends SdsAbstractExpression {
	leftOperand?: SdsAbstractExpression
	operator?: string
	rightOperand?: SdsAbstractExpression
}

interface SdsPrefixOperation extends SdsAbstractExpression {
	operand?: SdsAbstractExpression
	operator?: string
}

interface SdsCall extends SdsAbstractCall, SdsAbstractChainedExpression {
	typeArgumentList?: SdsTypeArgumentList
}

interface SdsAbstractChainedExpression extends SdsAbstractExpression {
	receiver?: SdsAbstractExpression
}

interface SdsIndexedAccess extends SdsAbstractChainedExpression {
	index?: SdsAbstractExpression
}

interface SdsMemberAccess extends SdsAbstractChainedExpression {
	member?: SdsReference
	nullSafe: boolean
}

interface SdsReference extends SdsAbstractExpression {
	declaration?: @SdsAbstractDeclaration
}

interface SdsSchemaReference extends SdsAbstractExpression {
	^type?: SdsSchemaType
}

interface SdsBoolean extends SdsAbstractLiteral {
	true: boolean
}

interface SdsFloat extends SdsAbstractNumber {
	value?: number
}

interface SdsAbstractNumber extends SdsAbstractLiteral {
}

interface SdsInt extends SdsAbstractNumber {
	value?: number
}

interface SdsNull extends SdsAbstractLiteral {
}

interface SdsParenthesizedExpression extends SdsAbstractExpression {
	expression?: SdsAbstractExpression
}

interface SdsTemplateString extends SdsAbstractExpression {
	expressions: SdsAbstractExpression[]
}

interface SdsTemplateStringStart extends SdsAbstractTemplateStringPart {
}

interface SdsAbstractTemplateStringPart extends SdsAbstractLiteral {
	value?: string
}

interface SdsTemplateStringInner extends SdsAbstractTemplateStringPart {
}

interface SdsTemplateStringEnd extends SdsAbstractTemplateStringPart {
}

interface SdsCallableType extends SdsAbstractCallable, SdsAbstractType {
	resultList?: SdsResultList
}

interface SdsUnionType extends SdsAbstractType {
	typeArgumentList?: SdsTypeArgumentList
}

interface SdsTypeProjection extends SdsAbstractTypeArgumentValue {
	^type?: SdsAbstractType
	variance?: string
}

interface SdsTypeParameterConstraint extends SdsAbstractStatement {
	leftOperand?: @SdsTypeParameter
	operator?: string
	rightOperand?: SdsAbstractType
}

interface SdsStarProjection extends SdsAbstractTypeArgumentValue {
}

