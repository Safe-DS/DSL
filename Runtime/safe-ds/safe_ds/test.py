import numpy as np
import pandas as pd

from safe_ds.data import Table, TableSchema, Row

table = Table.from_json("../tests/resources/test_schema_table.json")
rows_is: list[Row] = table.to_rows()
table_is: Table = Table.from_rows(rows_is)

print(table._data)
print("\n")
print(table_is._data)
print("\n")
for (index, series_row) in table._data.iterrows():
    print(series_row)
print("\n")
for row in rows_is:
    print(row._data)
