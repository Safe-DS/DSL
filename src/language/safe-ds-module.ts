import {
    createDefaultModule,
    createDefaultSharedModule,
    DeepPartial,
    DefaultSharedModuleContext,
    inject,
    LangiumServices,
    LangiumSharedServices,
    Module,
    PartialLangiumServices,
} from 'langium';
import { SafeDsGeneratedModule, SafeDsGeneratedSharedModule } from './generated/module.js';
import { registerValidationChecks, SafeDsValidator } from './validation/safe-ds-validator.js';
import { SafeDsFormatter } from './formatting/safe-ds-formatter.js';
import { SafeDsWorkspaceManager } from './builtins/safe-ds-workspace-manager.js';
import { SafeDsScopeComputation } from './scoping/safe-ds-scope-computation.js';
import { SafeDsScopeProvider } from './scoping/safe-ds-scope-provider.js';

/**
 * Declaration of custom services - add your own service classes here.
 */
export type SafeDsAddedServices = {
    validation: {
        SafeDsValidator: SafeDsValidator;
    };
};

/**
 * Union of Langium default services and your custom services - use this as constructor parameter
 * of custom service classes.
 */
export type SafeDsServices = LangiumServices & SafeDsAddedServices;

/**
 * Dependency injection module that overrides Langium default services and contributes the
 * declared custom services. The Langium defaults can be partially specified to override only
 * selected services, while the custom services must be fully specified.
 */
export const SafeDsModule: Module<SafeDsServices, PartialLangiumServices & SafeDsAddedServices> = {
    lsp: {
        Formatter: () => new SafeDsFormatter(),
    },
    references: {
        ScopeComputation: (services) => new SafeDsScopeComputation(services),
        ScopeProvider: (services) => new SafeDsScopeProvider(services),
    },
    validation: {
        SafeDsValidator: () => new SafeDsValidator(),
    },
};

export type SafeDsSharedServices = LangiumSharedServices;

export const SafeDsSharedModule: Module<SafeDsSharedServices, DeepPartial<SafeDsSharedServices>> = {
    workspace: {
        WorkspaceManager: (services) => new SafeDsWorkspaceManager(services),
    },
};

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection
 * @return An object wrapping the shared services and the language-specific services
 */
export const createSafeDsServices = function (context: DefaultSharedModuleContext): {
    shared: LangiumSharedServices;
    SafeDs: SafeDsServices;
} {
    const shared = inject(createDefaultSharedModule(context), SafeDsGeneratedSharedModule, SafeDsSharedModule);
    const SafeDs = inject(createDefaultModule({ shared }), SafeDsGeneratedModule, SafeDsModule);
    shared.ServiceRegistry.register(SafeDs);
    registerValidationChecks(SafeDs);
    return { shared, SafeDs };
};
