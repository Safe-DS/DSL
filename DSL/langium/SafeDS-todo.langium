grammar SafeDS
import 'SafeDS-types'




SdsAnnotationCall returns SdsAnnotationCall:
    '@' annotation=[SdsAnnotation:ID ] argumentList=SdsAnnotationCallArgumentList ?
;

SdsAnnotationCallArgumentList returns SdsArgumentList:
    {SdsArgumentList} '(' (arguments+=SdsAnnotationCallArgument  (',' arguments+=SdsAnnotationCallArgument  )* ','? )? ')'
;

SdsAnnotationCallArgument returns SdsArgument:
    (parameter=[SdsParameter:ID ] '=' )? value=SdsExpression
;

SdsParentTypeList returns SdsParentTypeList:
    'sub' parentTypes+=SdsParentType  (',' parentTypes+=SdsParentType  )* ','?
;



SdsClassBody returns SdsClassBody:
    {SdsClassBody} '{' members+=SdsClassMember * '}'
;

SdsClassMember returns SdsAbstractObject:
    SdsAnnotatedClassMember | SdsProtocol | SdsConstraint
;

SdsAnnotatedClassMember returns SdsAbstractAnnotatedObject:
    {SdsAnnotationCallList} annotationCalls+=SdsAnnotationCall * ({SdsAttribute.annotationCallList=current} static?='static' ? 'attr' name=ID  (':' ^type=SdsType  )?  | {SdsClass.annotationCallList=current} 'class' name=ID  typeParameterList=SdsTypeParameterList ? parameterList=SdsParameterList ? parentTypeList=SdsParentTypeList ? body=SdsClassBody ?  | {SdsEnum.annotationCallList=current} 'enum' name=ID  body=SdsEnumBody ?  | {SdsFunction.annotationCallList=current} static?='static' ? 'fun' name=ID  typeParameterList=SdsTypeParameterList ? parameterList=SdsParameterList  resultList=SdsResultList ? body=SdsFunctionBody ?  )
;

SdsConstraint returns SdsConstraint:
    'constraint' body=SdsConstraintBlock
;



SdsEnumBody returns SdsEnumBody:
    {SdsEnumBody} '{' variants+=SdsEnumVariant * '}'
;

SdsEnumVariant returns SdsEnumVariant:
    annotationCalls+=SdsAnnotationCall * name=ID  typeParameterList=SdsTypeParameterList ? parameterList=SdsParameterList ? constraint=SdsConstraint ?
;


SdsFunctionBody returns SdsFunctionBody:
    {SdsFunctionBody} '{' statements+=SdsFunctionStatement * '}'
;

SdsFunctionStatement returns SdsAbstractObject:
    SdsConstraint
;






SdsConstraintBlock returns SdsBlock:
    {SdsBlock} '{' (statements+=SdsConstraintStatement  (',' statements+=SdsConstraintStatement  )* )? '}'
;

SdsConstraintStatement returns SdsAbstractStatement:
    SdsPredicateAssignment | SdsPredicateExpressionStatement | SdsTypeParameterConstraint
;





terminal FLOAT returns number:DECIMAL_DIGIT+ '.' DECIMAL_DIGIT+ FLOAT_EXPONENT?  | DECIMAL_DIGIT+ FLOAT_EXPONENT  ;
terminal DECIMAL_DIGIT returns string:'0' ..'9' ;
terminal FLOAT_EXPONENT returns string:('e' | 'E' )('+' | '-' )?DECIMAL_DIGIT+  ;
terminal SCHEMA_EFFECT returns string:'$readSchema' | '$checkColumn' | '$removeColumn' | '$keepColumn' | '$renameColumn' | '$addColumn' | '$changeColumnType' ;
terminal INT returns number:DECIMAL_DIGIT+ ;
terminal STRING returns string:STRING_START STRING_TEXT* STRING_END  ;
terminal STRING_START returns string:STRING_DELIMITER ;
terminal STRING_END returns string:'{'? STRING_DELIMITER  ;
terminal STRING_DELIMITER returns string:'"' ;
terminal STRING_TEXT returns string:'{'? ESCAPE_SEQUENCE  | '{'?  !('\\' | STRING_DELIMITER | '{' ) ;
terminal ESCAPE_SEQUENCE returns string:'\\' . ;
terminal TEMPLATE_EXPRESSION_START returns string:'{{' ;
terminal TEMPLATE_EXPRESSION_END returns string:'}}' ;
terminal TEMPLATE_STRING_START returns string:STRING_START STRING_TEXT* TEMPLATE_EXPRESSION_START  ;
terminal TEMPLATE_STRING_INNER returns string:TEMPLATE_EXPRESSION_END STRING_TEXT* TEMPLATE_EXPRESSION_START  ;
terminal TEMPLATE_STRING_END returns string:TEMPLATE_EXPRESSION_END STRING_TEXT* STRING_END  ;
