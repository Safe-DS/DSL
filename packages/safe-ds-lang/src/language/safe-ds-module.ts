import { DeepPartial, inject, Module } from 'langium';
import {
    createDefaultModule,
    createDefaultSharedModule,
    DefaultSharedModuleContext,
    LangiumServices,
    LangiumSharedServices,
    PartialLangiumServices,
} from 'langium/lsp';
import { SafeDsAnnotations } from './builtins/safe-ds-annotations.js';
import { SafeDsClasses } from './builtins/safe-ds-classes.js';
import { SafeDsEnums, SafeDsImpurityReasons } from './builtins/safe-ds-enums.js';
import { SafeDsCommentProvider } from './documentation/safe-ds-comment-provider.js';
import { SafeDsDocumentationProvider } from './documentation/safe-ds-documentation-provider.js';
import { SafeDsCallGraphComputer } from './flow/safe-ds-call-graph-computer.js';
import { SafeDsGeneratedModule, SafeDsGeneratedSharedModule } from './generated/module.js';
import { SafeDsPythonGenerator } from './generation/safe-ds-python-generator.js';
import { SafeDsValueConverter } from './grammar/safe-ds-value-converter.js';
import { SafeDsNodeMapper } from './helpers/safe-ds-node-mapper.js';
import { SafeDsCallHierarchyProvider } from './lsp/safe-ds-call-hierarchy-provider.js';
import { SafeDsDocumentSymbolProvider } from './lsp/safe-ds-document-symbol-provider.js';
import { SafeDsFormatter } from './lsp/safe-ds-formatter.js';
import { SafeDsInlayHintProvider } from './lsp/safe-ds-inlay-hint-provider.js';
import { SafeDsNodeInfoProvider } from './lsp/safe-ds-node-info-provider.js';
import { SafeDsNodeKindProvider } from './lsp/safe-ds-node-kind-provider.js';
import { SafeDsSemanticTokenProvider } from './lsp/safe-ds-semantic-token-provider.js';
import { SafeDsSignatureHelpProvider } from './lsp/safe-ds-signature-help-provider.js';
import { SafeDsTypeHierarchyProvider } from './lsp/safe-ds-type-hierarchy-provider.js';
import { SafeDsPartialEvaluator } from './partialEvaluation/safe-ds-partial-evaluator.js';
import { SafeDsScopeComputation } from './scoping/safe-ds-scope-computation.js';
import { SafeDsScopeProvider } from './scoping/safe-ds-scope-provider.js';
import { SafeDsClassHierarchy } from './typing/safe-ds-class-hierarchy.js';
import { SafeDsCoreTypes } from './typing/safe-ds-core-types.js';
import { SafeDsTypeChecker } from './typing/safe-ds-type-checker.js';
import { SafeDsTypeComputer } from './typing/safe-ds-type-computer.js';
import { registerValidationChecks } from './validation/safe-ds-validator.js';
import { SafeDsDocumentBuilder } from './workspace/safe-ds-document-builder.js';
import { SafeDsPackageManager } from './workspace/safe-ds-package-manager.js';
import { SafeDsWorkspaceManager } from './workspace/safe-ds-workspace-manager.js';
import { SafeDsPurityComputer } from './purity/safe-ds-purity-computer.js';
import { SafeDsSettingsProvider } from './workspace/safe-ds-settings-provider.js';
import { SafeDsRenameProvider } from './lsp/safe-ds-rename-provider.js';
import { SafeDsRunner } from './runner/safe-ds-runner.js';
import { SafeDsTypeFactory } from './typing/safe-ds-type-factory.js';

/**
 * Declaration of custom services - add your own service classes here.
 */
export type SafeDsAddedServices = {
    builtins: {
        Annotations: SafeDsAnnotations;
        Classes: SafeDsClasses;
        Enums: SafeDsEnums;
        ImpurityReasons: SafeDsImpurityReasons;
    };
    evaluation: {
        PartialEvaluator: SafeDsPartialEvaluator;
    };
    flow: {
        CallGraphComputer: SafeDsCallGraphComputer;
    };
    generation: {
        PythonGenerator: SafeDsPythonGenerator;
    };
    helpers: {
        NodeMapper: SafeDsNodeMapper;
    };
    lsp: {
        NodeInfoProvider: SafeDsNodeInfoProvider;
    };
    purity: {
        PurityComputer: SafeDsPurityComputer;
    };
    types: {
        ClassHierarchy: SafeDsClassHierarchy;
        CoreTypes: SafeDsCoreTypes;
        TypeChecker: SafeDsTypeChecker;
        TypeComputer: SafeDsTypeComputer;
        TypeFactory: SafeDsTypeFactory;
    };
    workspace: {
        PackageManager: SafeDsPackageManager;
        SettingsProvider: SafeDsSettingsProvider;
    };
    runtime: {
        Runner: SafeDsRunner;
    };
};

/**
 * Union of Langium default services and your custom services - use this as constructor parameter
 * of custom service classes.
 */
export type SafeDsServices = LangiumServices & SafeDsAddedServices;

/**
 * Dependency injection module that overrides Langium default services and contributes the
 * declared custom services. The Langium defaults can be partially specified to override only
 * selected services, while the custom services must be fully specified.
 */
export const SafeDsModule: Module<SafeDsServices, PartialLangiumServices & SafeDsAddedServices> = {
    builtins: {
        Annotations: (services) => new SafeDsAnnotations(services),
        Classes: (services) => new SafeDsClasses(services),
        Enums: (services) => new SafeDsEnums(services),
        ImpurityReasons: (services) => new SafeDsImpurityReasons(services),
    },
    documentation: {
        CommentProvider: (services) => new SafeDsCommentProvider(services),
        DocumentationProvider: (services) => new SafeDsDocumentationProvider(services),
    },
    evaluation: {
        PartialEvaluator: (services) => new SafeDsPartialEvaluator(services),
    },
    flow: {
        CallGraphComputer: (services) => new SafeDsCallGraphComputer(services),
    },
    generation: {
        PythonGenerator: (services) => new SafeDsPythonGenerator(services),
    },
    helpers: {
        NodeMapper: (services) => new SafeDsNodeMapper(services),
    },
    lsp: {
        CallHierarchyProvider: (services) => new SafeDsCallHierarchyProvider(services),
        DocumentSymbolProvider: (services) => new SafeDsDocumentSymbolProvider(services),
        Formatter: () => new SafeDsFormatter(),
        InlayHintProvider: (services) => new SafeDsInlayHintProvider(services),
        NodeInfoProvider: (services) => new SafeDsNodeInfoProvider(services),
        RenameProvider: (services) => new SafeDsRenameProvider(services),
        SemanticTokenProvider: (services) => new SafeDsSemanticTokenProvider(services),
        SignatureHelp: (services) => new SafeDsSignatureHelpProvider(services),
        TypeHierarchyProvider: (services) => new SafeDsTypeHierarchyProvider(services),
    },
    parser: {
        ValueConverter: () => new SafeDsValueConverter(),
    },
    purity: {
        PurityComputer: (services) => new SafeDsPurityComputer(services),
    },
    references: {
        ScopeComputation: (services) => new SafeDsScopeComputation(services),
        ScopeProvider: (services) => new SafeDsScopeProvider(services),
    },
    types: {
        ClassHierarchy: (services) => new SafeDsClassHierarchy(services),
        CoreTypes: (services) => new SafeDsCoreTypes(services),
        TypeChecker: (services) => new SafeDsTypeChecker(services),
        TypeComputer: (services) => new SafeDsTypeComputer(services),
        TypeFactory: (services) => new SafeDsTypeFactory(services),
    },
    workspace: {
        PackageManager: (services) => new SafeDsPackageManager(services),
        SettingsProvider: (services) => new SafeDsSettingsProvider(services),
    },
    runtime: {
        Runner: (services) => new SafeDsRunner(services),
    },
};

export type SafeDsSharedServices = LangiumSharedServices;

export const SafeDsSharedModule: Module<SafeDsSharedServices, DeepPartial<SafeDsSharedServices>> = {
    lsp: {
        NodeKindProvider: () => new SafeDsNodeKindProvider(),
    },
    workspace: {
        DocumentBuilder: (services) => new SafeDsDocumentBuilder(services),
        WorkspaceManager: (services) => new SafeDsWorkspaceManager(services),
    },
};

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection.
 * @param options Further options to configure the Safe-DS module.
 * @return An object wrapping the shared services and the language-specific services.
 */
export const createSafeDsServices = function (
    context: DefaultSharedModuleContext,
    options?: ModuleOptions,
): {
    shared: LangiumSharedServices;
    SafeDs: SafeDsServices;
} {
    const shared = inject(createDefaultSharedModule(context), SafeDsGeneratedSharedModule, SafeDsSharedModule);
    const SafeDs = inject(createDefaultModule({ shared }), SafeDsGeneratedModule, SafeDsModule);
    shared.ServiceRegistry.register(SafeDs);
    registerValidationChecks(SafeDs);
    SafeDs.runtime.Runner.updateRunnerCommand(options?.runnerCommand);
    return { shared, SafeDs };
};

/**
 * Create the full set of services required by Langium and load all builtin files into the workspace.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection.
 * @param options Further options to configure the Safe-DS module.
 * @return An object wrapping the shared services and the language-specific services.
 */
export const createSafeDsServicesWithBuiltins = async function (
    context: DefaultSharedModuleContext,
    options?: ModuleOptions,
): Promise<{
    shared: LangiumSharedServices;
    SafeDs: SafeDsServices;
}> {
    const { shared, SafeDs } = createSafeDsServices(context, options);
    await shared.workspace.WorkspaceManager.initializeWorkspace([]);
    return { shared, SafeDs };
};

/**
 * Options to pass to the creation of Safe-DS services.
 */
export interface ModuleOptions {
    /**
     * Optional command to start the runner, if available.
     */
    runnerCommand?: string;
}
