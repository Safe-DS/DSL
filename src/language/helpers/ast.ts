import { AstNode, hasContainerOfType } from 'langium';

/**
 * Returns whether the inner node is contained in the outer node. If the nodes are equal, this function returns `true`.
 */
export const isContainedIn = (inner: AstNode | undefined, outer: AstNode | undefined): boolean => {
    return hasContainerOfType(inner, (it) => it === outer);
};
