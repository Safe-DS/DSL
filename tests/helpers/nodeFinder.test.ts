import { afterEach, describe, expect, it } from 'vitest';
import { getFirstNodeOfType, getNodeByLocation } from './nodeFinder.js';
import { createSafeDsServices } from '../../src/language/safe-ds-module.js';
import { EmptyFileSystem } from 'langium';
import { AssertionError } from 'assert';
import { clearDocuments, parseHelper } from 'langium/test';
import { isSdsClass } from '../../src/language/generated/ast.js';

describe('getNodeByLocation', () => {
    const services = createSafeDsServices(EmptyFileSystem).SafeDs;

    afterEach(async () => {
        await clearDocuments(services);
    });

    it('should throw if no document is found', () => {
        expect(() => {
            getNodeByLocation(services, {
                uri: 'file:///test.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            });
        }).toThrowErrorMatchingSnapshot();
    });

    it('should throw if no node is found', async () => {
        const document = await parseHelper(services)(`class C`);

        expect(() => {
            getNodeByLocation(services, {
                uri: document.uri.toString(),
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            });
        }).toThrow(AssertionError);
    });

    it('should return the node that fills the range completely', async () => {
        const document = await parseHelper(services)(`class C`);

        expect(
            getNodeByLocation(services, {
                uri: document.uri.toString(),
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 7 } },
            }),
        ).to.satisfy(isSdsClass);
    });

    it('should return the node whose name fills the range completely', async () => {
        const document = await parseHelper(services)(`class C`);

        expect(
            getNodeByLocation(services, {
                uri: document.uri.toString(),
                range: { start: { line: 0, character: 6 }, end: { line: 0, character: 7 } },
            }),
        ).to.satisfy(isSdsClass);
    });
});

describe('getFirstNodeOfType', () => {
    const services = createSafeDsServices(EmptyFileSystem).SafeDs;

    afterEach(async () => {
        await clearDocuments(services);
    });

    it('should throw if no node is found', async () => {
        const code = '';
        expect(async () => {
            await getFirstNodeOfType(services, code, isSdsClass);
        }).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should return the first matching node', async () => {
        const code = 'class C';
        const node = await getFirstNodeOfType(services, code, isSdsClass);
        expect(node).to.satisfy(isSdsClass);
    });
});
