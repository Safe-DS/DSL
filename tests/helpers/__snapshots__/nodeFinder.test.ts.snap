// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getFirstNodeOfType > should throw if no node is found 1`] = `"Expected to find a matching node but found none."`;

exports[`getNodeByLocation > should throw if no document is found 1`] = `"No document found at file:///test.sdstest"`;
